drivers/adc/adc_sam_afec.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2017 comsuisse AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief Atmel SAM MCU family ADC (AFEC) driver.
+ */
+
+#include <errno.h>
+#include <misc/__assert.h>
+#include <misc/util.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <adc.h>
+
+#define SYS_LOG_DOMAIN "dev/adc_sam_afec"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_ADC_LEVEL
+#include <logging/sys_log.h>
+
+#define ADC_CHANNELS 12
+
+struct adc_sam_dev_cfg {
+	Afec *regs;
+	const struct soc_gpio_pin *pin_trigger;
+	void (*irq_config)(void);
+	u8_t irq_id;
+	u8_t periph_id;
+};
+
+struct channel_samples {
+	u16_t *buffer;
+	u32_t length;
+};
+
+struct adc_sam_dev_data {
+	struct k_sem sem_meas;
+	struct k_sem mutex_thread;
+	u16_t active_channels;
+	u16_t measured_channels;
+	u16_t active_chan_last;
+	struct channel_samples samples[ADC_CHANNELS];
+};
+
+#define CONF_ADC_PRESCALER ((SOC_ATMEL_SAM_MCK_FREQ_HZ / 15000000) - 1)
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_CFG(dev) \
+	((const struct adc_sam_dev_cfg *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct adc_sam_dev_data *const)(dev)->driver_data)
+
+static void adc_sam_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct adc_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct adc_sam_dev_data *const dev_data = DEV_DATA(dev);
+	Afec *const afec = dev_cfg->regs;
+	struct channel_samples *samples = dev_data->samples;
+	u16_t isr_status;
```

Comment:
isr_status most likely will be implemented by compiler as a register variable, in this case using u32_t is slightly more efficient (one assembly instruction less).